src/compiler/checker.rs

```diff
@@ -170,6 +182,109 @@ impl TypeChecker {
         type_
     }
 
+    fn type_to_string(&self, type_: &Type) -> String {
+        unimplemented!()
+    }
+
+    fn get_type_names_for_error_display(&self, left: &Type, right: &Type) -> (String, String) {
+        let left_str = if false {
+            unimplemented!()
+        } else {
+            self.type_to_string(left)
+        };
+        let right_str = if false {
+            unimplemented!()
+        } else {
+            self.type_to_string(right)
+        };
+        (left_str, right_str)
+    }
+
+    fn get_type_name_for_error_display(&self, type_: &Type) -> String {
+        self.type_to_string(type_)
+    }
+
+    fn add_optionality(&self, type_: Rc<Type>) -> Rc<Type> {
+        type_
+    }
+
+    fn get_type_for_variable_like_declaration(&self, declaration: &Node) -> Option<Rc<Type>> {
+        let declared_type = self.try_get_type_from_effective_type_node(declaration);
+        if let Some(declared_type) = declared_type {
+            return Some(self.add_optionality(declared_type));
+        }
+        unimplemented!()
+    }
+
+    fn get_widened_type_for_variable_like_declaration(&self, declaration: &Node) -> Rc<Type> {
+        self.widen_type_for_variable_like_declaration(
+            self.get_type_for_variable_like_declaration(declaration),
+            declaration,
+        )
+    }
+
+    fn widen_type_for_variable_like_declaration(
+        &self,
+        type_: Option<Rc<Type>>,
+        declaration: &Node,
+    ) -> Rc<Type> {
+        if let Some(type_) = type_ {
+            return self.get_widened_type(type_);
+        }
+        unimplemented!()
+    }
+
+    fn try_get_type_from_effective_type_node(
+        &self,
+        declaration: &Node, /*Declaration*/
+    ) -> Option<Rc<Type>> {
+        let type_node = get_effective_type_annotation_node(declaration);
+        type_node.map(|type_node| self.get_type_from_type_node(&*type_node))
+    }
```

Comment:
I don't know if it's more clear, but you could write this function using the `?` option syntax:

```rs
fn try_get_type_from_effective_type_node(
    &self,
    declaration: &Node, /*Declaration*/
) -> Option<Rc<Type>> {
    let type_node = get_effective_type_annotation_node(declaration)?;
    Some(self.get_type_from_type_node(&*type_node))
}
```

Reply:
Ah interesting, I forgot that you can use `?` for `Option`'s in addition to `Result`'s